src/task_queue.rs

```diff
@@ -0,0 +1,326 @@
+
+  use crate::JoinSet;
+
+  use super::*;
+
+  #[tokio::test]
+  async fn task_queue_runs_one_after_other() {
+    let task_queue = Rc::new(TaskQueue::default());
+    let mut set = JoinSet::default();
+    let data = Arc::new(Mutex::new(0));
+    for i in 0..100 {
+      let data = data.clone();
+      let task_queue = task_queue.clone();
+      let acquire = task_queue.acquire();
+      set.spawn(async move {
+        let permit = acquire.await;
+        crate::spawn_blocking(move || {
+          let mut data = data.lock().unwrap();
+          assert_eq!(i, *data);
+          *data = i + 1;
+        })
+        .await
+        .unwrap();
+        drop(permit);
+        drop(task_queue);
+      });
+    }
+    while let Some(res) = set.join_next().await {
+      assert!(res.is_ok());
+    }
+  }
+
+  #[tokio::test]
+  async fn tasks_run_in_sequence() {
+    let task_queue = Rc::new(TaskQueue::default());
+    let data = RefCell::new(0);
+
+    let first = task_queue.run(async {
+      *data.borrow_mut() = 1;
+    });
+    let second = task_queue.run(async {
+      assert_eq!(*data.borrow(), 1);
+      *data.borrow_mut() = 2;
+    });
+    let _ = tokio::join!(first, second);
+
+    assert_eq!(*data.borrow(), 2);
+  }
+
+  #[tokio::test]
+  async fn future_dropped_before_poll() {
+    let task_queue = Rc::new(TaskQueue::default());
+
+    // acquire a future, but do not await it
+    let future = task_queue.acquire();
+
+    // this task tries to acquire another permit, but will be blocked by the first permit.
+    let flag = Rc::new(Flag::default());
+    let delayed_task = crate::spawn({
+      let task_queue = task_queue.clone();
+      let flag = flag.clone();
+      async move {
+        flag.raise();
+        task_queue.acquire().await;
+        true
+      }
+    });
+
+    // ensure the task gets a chance to be scheduled and blocked
+    tokio::task::yield_now().await;
+    assert!(flag.is_raised());
+
+    // now, drop the first future
+    drop(future);
+
+    assert!(delayed_task.await.unwrap());
+  }
+
+  #[tokio::test]
+  async fn many_future_dropped_before_poll() {
+    let task_queue = Rc::new(TaskQueue::default());
+
+    // acquire a future, but do not await it
+    let mut futures = Vec::new();
+    for _ in 0..5 {
+      futures.push(task_queue.acquire());
+    }
+
+    // this task tries to acquire another permit, but will be blocked by the first permit.
+    let flag = Rc::new(Flag::default());
+    let delayed_task = crate::spawn({
+      let task_queue = task_queue.clone();
+      let flag = flag.clone();
+      async move {
+        flag.raise();
+        task_queue.acquire().await;
+        true
+      }
+    });
+
+    // ensure the task gets a chance to be scheduled and blocked
+    tokio::task::yield_now().await;
+    assert!(flag.is_raised());
+
+    // now, drop the futures
+    drop(futures);
+
+    assert!(delayed_task.await.unwrap());
+  }
+
+  #[tokio::test]
+  async fn acquires_position_synchronously() {
```

Comment:
Unlike the previous implementation, this implementation allows acquiring the position synchronously.